File: apps-script/ConnectorOllama.js
// ================================================================
// 📚 Connector Ollama — Bibliothèque Générale PreCogn
// ================================================================
//
// Relais vers le service Ollama local via Analyzor.
// Aucune logique métier — API uniforme uniquement.
//
// Utilisé par : LLMPreCogn, Facilitator, Agents, Flows
// ================================================================

/**
 * Vérifie que le service Ollama répond.
 * @returns {{ success: boolean, status?: string, error?: string }}
 */
function ollamaStatus() {
  return analyzor("/api/ollama/status", null, "GET");
}

/**
 * Liste les modèles installés sur Ollama.
 * @returns {{ success: boolean, models?: Object[], error?: string }}
 */
function ollamaModels() {
  return analyzor("/api/ollama/models", null, "GET");
}

/**
 * Génère une complétion à partir d'un prompt (API /generate).
 * @param {string}  prompt
 * @param {string=} model    Défaut côté serveur : qwen2.5-coder:3b
 * @param {Object=} options  Paramètres Ollama (temperature, num_ctx, ...)
 * @returns {{ success: boolean, content?: string, model?: string, error?: string }}
 */
function ollamaGenerate(prompt, model, options) {
  const payload = { prompt: prompt };
  if (model)   payload.model   = model;
  if (options) payload.options = options;
  return analyzor("/api/ollama/generate", payload, "POST");
}

/**
 * Conversation multi-tour (API /chat).
 * @param {Object[]} messages  [{role, content}, ...]
 * @param {string=}  model     Défaut côté serveur : qwen2.5-coder:3b
 * @param {Object=}  options   Paramètres Ollama (temperature, num_ctx, ...)
 * @returns {{ success: boolean, content?: string, model?: string, error?: string }}
 */
function ollamaChat(messages, model, options) {
  const payload = { messages: messages };
  if (model)   payload.model   = model;
  if (options) payload.options = options;
  return analyzor("/api/ollama/chat", payload, "POST");
}

/**
 * Génère un embedding vectoriel.
 * @param {string}  text
 * @param {string=} model  Défaut côté serveur : nomic-embed-text
 * @returns {{ success: boolean, embedding?: number[], error?: string }}
 */
function ollamaEmbedding(text, model) {
  const payload = { input: text };
  if (model) payload.model = model;
  return analyzor("/api/ollama/embed", payload, "POST");
}

File: python/connector_ollama.py
"""
Connector Ollama — isole tout le reste du programme des appels HTTP vers
le service Ollama local (http://127.0.0.1:11434).

Si Ollama change d'API ou est remplacé par un autre moteur local,
seul ce fichier doit changer.

Interface stable exposée :
- status()          -> dict
- models()          -> dict
- generate(payload) -> dict
- chat(payload)     -> dict
- embed(payload)    -> dict
"""

import requests

OLLAMA_URL = "http://127.0.0.1:11434"
DEFAULT_CHAT_MODEL  = "qwen2.5-coder:3b"
DEFAULT_EMBED_MODEL = "nomic-embed-text"
TIMEOUT = 120  # secondes — les modèles locaux peuvent être lents au premier appel


def _post(endpoint: str, body: dict, timeout: int = TIMEOUT) -> dict:
    try:
        r = requests.post(f"{OLLAMA_URL}{endpoint}", json=body, timeout=timeout)
        r.raise_for_status()
        return r.json()
    except requests.Timeout:
        return {"success": False, "error": f"Ollama timeout ({TIMEOUT}s)"}
    except requests.ConnectionError:
        return {"success": False, "error": "Ollama non joignable (127.0.0.1:11434)"}
    except requests.HTTPError:
        return {"success": False, "error": f"Ollama HTTP {r.status_code}: {r.text[:200]}"}
    except Exception as e:
        return {"success": False, "error": str(e)}


def _get(endpoint: str) -> dict:
    try:
        r = requests.get(f"{OLLAMA_URL}{endpoint}", timeout=10)
        r.raise_for_status()
        return r.json()
    except requests.Timeout:
        return {"success": False, "error": "Ollama timeout"}
    except requests.ConnectionError:
        return {"success": False, "error": "Ollama non joignable (127.0.0.1:11434)"}
    except requests.HTTPError:
        return {"success": False, "error": f"Ollama HTTP {r.status_code}"}
    except Exception as e:
        return {"success": False, "error": str(e)}


def status() -> dict:
    """Vérifie que le service Ollama répond."""
    try:
        r = requests.get(OLLAMA_URL, timeout=5)
        return {"success": True, "status": r.text.strip()}
    except requests.ConnectionError:
        return {"success": False, "error": "Ollama non joignable (127.0.0.1:11434)"}
    except Exception as e:
        return {"success": False, "error": str(e)}


def models() -> dict:
    """Liste les modèles installés."""
    data = _get("/api/tags")
    if "success" in data:  # erreur remontée par _get
        return data
    return {"success": True, "models": data.get("models", [])}


def generate(payload: dict) -> dict:
    """Génère une complétion à partir d'un prompt (API /generate).

    payload attendu :
      - prompt  (str)  : obligatoire
      - model   (str)  : optionnel, défaut DEFAULT_CHAT_MODEL
      - options (dict) : optionnel
      - timeout (int)  : optionnel, défaut TIMEOUT (120s) — à réduire pour un appelant qui a
        un budget de temps strict (ex. Communicator, voir understand.py::_call_ollama, retour
        de Stéphane 2026-08-03 : "5 secondes max").
    """
    prompt = payload.get("prompt")
    if not prompt:
        return {"success": False, "error": "prompt requis"}

    model = payload.get("model") or DEFAULT_CHAT_MODEL
    body: dict = {"model": model, "prompt": prompt, "stream": False}
    if payload.get("options"):
        body["options"] = payload["options"]

    data = _post("/api/generate", body, timeout=payload.get("timeout", TIMEOUT))
    if "success" in data:
        return data
    return {"success": True, "content": data.get("response") or "", "model": data.get("model") or model}


def chat(payload: dict) -> dict:
    """Conversation multi-tour (API /chat).

    payload attendu :
      - messages (list[dict]) : obligatoire, format [{role, content}, ...]
      - model    (str)        : optionnel, défaut DEFAULT_CHAT_MODEL
      - options  (dict)       : optionnel
    """
    messages = payload.get("messages")
    if not messages:
        return {"success": False, "error": "messages requis"}

    model = payload.get("model") or DEFAULT_CHAT_MODEL
    body: dict = {"model": model, "messages": messages, "stream": False}
    if payload.get("options"):
        body["options"] = payload["options"]

    data = _post("/api/chat", body)
    if "success" in data:
        return data
    content = (data.get("message") or {}).get("content") or ""
    return {"success": True, "content": content, "model": data.get("model") or model}


def embed(payload: dict) -> dict:
    """Génère un embedding vectoriel (API /api/embed).

    payload attendu :
      - input (str)  : obligatoire
      - model (str)  : optionnel, défaut DEFAULT_EMBED_MODEL
    """
    text = payload.get("input")
    if not text:
        return {"success": False, "error": "input requis"}

    model = payload.get("model") or DEFAULT_EMBED_MODEL
    data = _post("/api/embed", {"model": model, "input": text})
    if "success" in data:
        return data
    return {"success": True, "embedding": data.get("embeddings", [data.get("embedding")])[0], "model": model}
